### PR TITLE
Refactor/psd 2555 file upload component functionality

### PR DESCRIPTION
**Pull request by tahir-khalid**
## Description
1. Updates the File Upload control to always display
2. Handled invalid image type with page validation
3. Adds test for invalid image type

## Screen-shots or screen-capture of UI changes
Invalid image type validation message (hyperlinked text that takes you to the file upload control)
image

File upload control demonstrating UI changes 
image

## Review apps

https://psd-pr-xxxx.london.cloudapps.digital/
https://psd-pr-xxxx-support.london.cloudapps.digital/
https://psd-pr-xxxx-report.london.cloudapps.digital/

## Checklist:
- [x] Have you documented your changes in the pull request description?
- [ ] Does the change present any security considerations?
- [ ] Is any gem functionality overloaded? Eg: Devise controller methods being overloaded.
- [ ] Has acceptance criteria been tested by a peer?

### General testing (author)
- [x] Test without JavaScript
- [x] Test on small screen

### Accessibility testing (author)
- [x] Reviewed by Designer (if required)
- [x] Works keyboard only
- [x] Tested with one screen reader
- [x] Zoom page to 400% - content still visible
- [x] Disable CSS - does content make sense and appear in a logical order?

### CODE REVIEW

File: app/forms/product_form.rb [hunk 1]
```diff
@@ -82,6 +83,14 @@ def cache_file!(user, product)
     end
   end
 
+  def acceptable_image
```
Comment (alan-at-work) [inline app/forms/product_form.rb:86]: Could make this a little better for example

```
ACCEPTABLE_IMAGE_TYPES = ["image/jpeg", "image/png", "image/gif", "image/webp", "image/heic", "image/heif"].freeze

def acceptable_image
  return if image.blank?

  unless ACCEPTABLE_IMAGE_TYPES.include?(image.content_type)
    errors.add(:image, "The selected file must be a JPG, PNG, GIF, WEBP, HEIC or HEIF")
  end
end
```

Could even add the error to the localisation file :) 

```
en:
  errors:
    messages:
      invalid_image_type: "The selected file must be a JPG, PNG, GIF, WEBP, HEIC or HEIF"
  ```
  
  Then change the code to read `errors.add(:image, I18n.t('errors.messages.invalid_image_type'))`

File: app/helpers/application_helper.rb [hunk 2]
```diff
@@ -50,15 +50,8 @@ def replace_uploaded_file_field(form, field_name, label:, label_classes: "govuk-
...
-    return safe_join([existing_uploaded_file_id, file_upload_field]) if uploaded_file.blank?
-
-    safe_join(
-      [
-        existing_uploaded_file_id,
-        render(partial: "active_storage/blobs/blob", locals: { blob: uploaded_file }),
-        govuk_details(summary_text: "Replace this file", text: file_upload_field)
-      ]
-    )
+    return existing_uploaded_file_id.to_s + file_upload_field.to_s + render(partial: "active_storage/blobs/blob", locals: { blob: uploaded_file }) if uploaded_file.present?
```
Comment (alan-at-work) [inline app/helpers/application_helper.rb:53]: could clean this up and separate the logic a bit maybe?

```
def replace_uploaded_file_field(form, field_name, label:, label_classes: "govuk-label--m")
  existing_uploaded_file_id = form.hidden_field "existing_#{field_name}_file_id"
  file_upload_field = form.govuk_file_upload(field_name, label: label, label_classes: label_classes)
  uploaded_file = form.object.public_send(field_name)

  return render_uploaded_file(existing_uploaded_file_id, file_upload_field, uploaded_file) if uploaded_file.present?

  "#{existing_uploaded_file_id}#{file_upload_field}"
end

private

def render_uploaded_file(existing_uploaded_file_id, file_upload_field, uploaded_file)
  "#{existing_uploaded_file_id}#{file_upload_field}" +
    render(partial: "active_storage/blobs/blob", locals: { blob: uploaded_file })
end
```

File: spec/features/add_product_spec.rb [hunk 3]
```diff
@@ -6,6 +6,9 @@
     attributes_for(:product_iphone, authenticity: Product.authenticities.keys.without("missing", "unsure").sample)
   end
 
+  let(:bad_file) { Rails.root.join "spec/fixtures/files/testImages.bin" }
```
Comment (alan-at-work) [inline spec/features/add_product_spec.rb:9]: To make these clearer on what's orrucing maybe use

```
let(:invalid_image_file) { Rails.root.join("spec/fixtures/files/testImages.bin") }
let(:valid_image_file) { Rails.root.join("spec/fixtures/files/testImage.png") }
```

File: spec/features/add_product_spec.rb [hunk 4]
```diff
@@ -181,7 +184,7 @@
     fill_in "Other product identifiers", with: attributes[:product_code]
     fill_in "Webpage", with: attributes[:webpage]
 
-    attach_file "product[image]", "spec/fixtures/files/testImage.png"
+    attach_file "product[image]", good_file
```
Comment (alan-at-work) [inline spec/features/add_product_spec.rb:187]: if you do the above this would need updating :)

**Review by alan-at-work — CHANGES_REQUESTED** [review]
Looks good just some points :) 

Would be cool to spend some time refactoring the spec. But we have a task for this in future

File: app/helpers/application_helper.rb (hunk 2 above)
Comment (tahir-khalid) [reply app/helpers/application_helper.rb:53]: Refactored.

File: spec/features/add_product_spec.rb (hunk 3 above)
Comment (tahir-khalid) [reply spec/features/add_product_spec.rb]: Refactored.

File: spec/features/add_product_spec.rb (hunk 4 above)
Comment (tahir-khalid) [reply spec/features/add_product_spec.rb:187]: Updated.

**Pull request closed without merging**